### PR TITLE
MSVC == 1900 scope_exit in lambda fix

### PR DESCRIPTION
**Pull request by tobias-loew**
fixed internal compiler error when using scope-exit inside a lambda with MSVC 1900 (VS 2015)

cf. http://boost.2283326.n4.nabble.com/scope-exit-Boost-gt-1-65-VS-2015-internal-compiler-error-in-lambda-td4700981.html

### CODE REVIEW

File: include/boost/typeof/typeof.hpp [hunk 1]
```diff
@@ -15,7 +15,7 @@
 
 #include <boost/config.hpp>
 
-#if !defined(BOOST_NO_CXX11_DECLTYPE) && !defined(BOOST_NO_CXX11_TEMPLATE_ALIASES) && !defined(BOOST_TYPEOF_EMULATION)
+#if (!defined(_MSC_VER) || (_MSC_VER > 1900)) && !defined(BOOST_NO_CXX11_DECLTYPE) && !defined(BOOST_NO_CXX11_TEMPLATE_ALIASES) && !defined(BOOST_TYPEOF_EMULATION)
```
Comment (Lastique) [inline include/boost/typeof/typeof.hpp]: Since this fix is specific to MSVC, it should test for `BOOST_MSVC`.

Comment (tobias-loew) [reply include/boost/typeof/typeof.hpp]: Done

**Review by jeking3 — CHANGES_REQUESTED** [review]
@pdimov wrote on the mailing list:
```
It will be very helpful if you couple this fix with a test
that demonstrates the breakage (that is, a test that fails
before the fix and passes after the fix.)
```

File: include/boost/typeof/typeof.hpp [hunk 2]
```diff
@@ -15,7 +15,7 @@
 
 #include <boost/config.hpp>
 
-#if !defined(BOOST_NO_CXX11_DECLTYPE) && !defined(BOOST_NO_CXX11_TEMPLATE_ALIASES) && !defined(BOOST_TYPEOF_EMULATION)
+#if (!defined(BOOST_MSVC) || (BOOST_MSVC > 1900)) && !defined(BOOST_NO_CXX11_DECLTYPE) && !defined(BOOST_NO_CXX11_TEMPLATE_ALIASES) && !defined(BOOST_TYPEOF_EMULATION)
```
Comment (pdimov) [inline include/boost/typeof/typeof.hpp]: Should probably use `&& !BOOST_WORKAROUND(BOOST_MSVC, <= 1900)` here, as that's what it is.

**Comment by tobias-loew** [comment]
I've made a pull-request for a test-case in boost/scope_exit since there the error happens. It is connected to boost::type_of::remove_cv_ref_t since the old implementation with boost::type_of::msvc_typeid_wrapper works. 
The following code when inside a lambda results in VS 2015 in an internal compiler error:

typedef void(*boost_se_tag_0_0)(int test);

       

        typedef boost::type_of::remove_cv_ref_t<

            decltype(::boost::scope_exit::detail::wrap(::boost::scope_exit::detail::deref(test, static_cast<boost_se_tag_0_0>(0))))

        > boost_se_wrapped_t_0_0;

       

        typedef  boost_se_wrapped_t_0_0::type boost_se_capture_t_0_0;

**Comment by pdimov** [comment]
The idea is to put the test case here in order to ensure that it won't be broken again; pull requests to `typeof` do not run the `scope_exit` tests.

File: include/boost/typeof/typeof.hpp [hunk 3]
```diff
@@ -15,7 +15,7 @@
 
 #include <boost/config.hpp>
```
Comment (pdimov) [inline include/boost/typeof/typeof.hpp]: Need to include `<bost/config/workaround.hpp>` here, I suppose.

**Comment by tobias-loew** [comment]
I've added a test case. It doesn't have any test assertions since the test is: either compiles or triggers internal compiler error.

**Comment by pdimov** [comment]
Thanks!

Unfortunately, this test fails to compile with `BOOST_TYPEOF_EMULATION` (the Jamfile executes each test with both `BOOST_TYPEOF_NATIVE` and `BOOST_TYPEOF_EMULATION`), and I've no idea why.

**Comment by pdimov** [comment]
I figured it out, `BOOST_TYPEOF_REGISTER_TEMPLATE(::detail::wrapper, 1)` was missing.

**Comment by pdimov** [comment]
Applied in https://github.com/boostorg/typeof/commit/080f05ea46271c745f7a748815a64210deb0a880.

**Comment by tobias-loew** [comment]
Great! Thanks!

**Pull request closed without merging**